WIP Modify REST framework to support deleting all resources in a namespace

kbeecher opened the pull request:
A WIP that so far aims to cover points 1 and 2 in @derekwaynecarr's comment: https://github.com/GoogleCloudPlatform/kubernetes/issues/5933#issuecomment-94466754

Refs: #5933

Commit:
Modify REST framework to support deleting all resources in a namespace

Refs: #5933

derekwaynecarr commented on pkg/api/rest/rest.go:
```diff
@@ -107,6 +107,13 @@ func (w GracefulDeleteAdapter) Delete(ctx api.Context, name string, options *api
 	return w.Deleter.Delete(ctx, name)
 }
 
+// An object that can delete a collection of RESTful resources in a namespace
```
Just say "An object can that can delete a collection of RESTful resources"

I could see this being implemented for non-namespaced resources, for example, to delete all nodes.

derekwaynecarr commented on pkg/api/rest/rest.go:
```diff
@@ -107,6 +107,13 @@ func (w GracefulDeleteAdapter) Delete(ctx api.Context, name string, options *api
+type CollectionDeleter interface {
+	// Deletes all resources of a certain type under a given namespace.
+	// It will return a status object containing information about the deletion.
+	DeleteCollection(namespace string) (runtime.Object, error)
```
Have this take a context object instead of a namespace

derekwaynecarr commented:
There was nothing immediately wrong with this, but when I went to test it on a vagrant setup, the kube-apiserver is failing to launch with the following output to log:

```
F0427 20:00:22.643514       1 server.go:244] Invalid Authentication Config: error reading /dev/null: could not read any certificates
```

This does not appear related to your PR, so I need to investigate that separately before I can test why you were not getting output to Swagger as you expected.

kbeecher commented:
@derekwaynecarr Understood. Thanks for taking the time.

kbeecher commented:
@derekwaynecarr I added a bit more code towards items 3 and 4.

Specifically, I added a "slice" of code that implements the other steps but only for pods. Hopefully this shows the basic outline of the approach (and asks a couple of questions too).

derekwaynecarr commented:
@kbeecher - I will dedicate time today to trying this out and giving feedback.  Thanks for your willingness to push the feature forward.

derekwaynecarr commented:
Tried this out, things look to be moving on the right path, swagger looked good, will review and comment further.

derekwaynecarr commented on pkg/api/rest/rest.go:
```diff
@@ -107,6 +107,13 @@ func (w GracefulDeleteAdapter) Delete(ctx api.Context, name string, options *api
 
+// An object that can delete a collection of RESTful resources.
+type CollectionDeleter interface {
+	// Deletes all resources of a certain type under a given namespace.
```
Deletes collection of resources  (remove the mention of namespaces entirely)

derekwaynecarr commented on pkg/client/pods.go:
```diff
@@ -73,6 +74,11 @@ func (c *pods) Delete(name string) error {
 
+// Removes all pods under the current namespace
+func (c *pods) DeleteAll() error {
+	return c.r.Delete().Namespace(c.ns).Resource("pods").LabelsSelectorParam(labels.Everything()).Do().Error()
```
This is the right general pattern, you do not need LabelsSelectorParam though

derekwaynecarr commented on pkg/registry/generic/etcd/etcd.go:
```diff
@@ -398,6 +398,11 @@ func (e *Etcd) Delete(ctx api.Context, name string, options *api.DeleteOptions)
 }
 
+// TODO Should this be here, or should each resource have it's own DeleteCollection()?
+func (e *Etcd) DeleteCollection(ctx api.Context) (runtime.Object, error) {
```
I think the behavior here is resource specific, so maybe let me pass a function to delegate the implementation of the desired behavior similar to what is done for the KeyFuncs, I suspect we will have a couple common functions.

For example, I see no issue for Events, Services, ReplicationControllers, ResourceQuotas, LimitRanges all supporting the ability to delete the entire collection even if it contains items - especially during namespace clean-up.  The function here would just do a etcd delete on KeyRootFunc, and things would just cascade nicely.

For Namespaces, I do not want to support DeleteCollection at this time, so I would want an implementation that results in the equivalent of a Forbidden response.

For Pods, I think we want to preserve the ability to do maintain support for graceful deletion.  As a result, I think pods would want a version of this function that would fail if the collection had any child nodes.  In practice, this means the namespace_controller would have to delete each pod in turn, and then once all pods are removed, then it would call DeleteCollection (since its empty now) to delete the root folder for pods in that namespace.  I would be interested in @smarterclayton take on this to see if I am being overly concerned.

For Nodes, I am not sure we want to support the operation at all.

derekwaynecarr replied:
Or for resources that want to support this pattern, apply an adapter that I can use to decorate the registry with this feature, and we would add a unique behavior at that point.

derekwaynecarr commented:
I like where this is headed, interested in getting some other feedback - most notably wrt to graceful delete.

smarterclayton commented:
Will have comments later on this, please hold.

> On Apr 28, 2015, at 11:39 AM, Derek Carr <notifications@github.com> wrote:
> 
> I like where this is headed, interested in getting some other feedback - most notably wrt to graceful delete.
> 
> —
> Reply to this email directly or view it on GitHub.
>

kbeecher commented:
@smarterclayton thanks for taking a look!

smarterclayton commented:
Approach looks fine to me, derek seems to have you covered on comments.

kbeecher commented:
@derekwaynecarr I'm going to resume work on this. Do you have any preference between the adapter/delegate approaches? Personally, I would choose the delegate approach, but that's only because I'm familiar with that, and not exactly sure right now how to implement an adapter :-)

smarterclayton commented:
A default delegate is probably better on generic.Etcd - do the default list implementation and have a number of selectable implementations an implementor can choose to use.

kbeecher commented:
I've added a little more, trying to implement according to your suggestions. If you guys think this is the right approach, I'll start 'widening' this patch to cover all other resources.

smarterclayton commented:
@csrwng can you describe the changes you are making as well so Karl is aware?

----- Original Message -----
> I've added a little more, trying to implement according to your suggestions.
> If you guys think this is the right approach, I'll start 'widening' this
> patch to cover all other resources.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/7372#issuecomment-99127981

csrwng commented:
@kbeecher @smarterclayton 
So basically we are saying that we will have a set of methods that act on a collection vs a set of methods that act on a named object. However, the naming still needs some thought (doesn't make sense to apply -Collection or -Named to all of them). Examples URIs are not necessarily real stuff:

| Method | Collection | Single Named Object | Sample URIs |
| ---------- | ------------- | ---------------------------- | ----------------- |
| GET | List | Get | Collection: */ns/pods*, One: */ns/pods/[name]*, Subresource: */ns/pods/[name]/log* |
| GET (w/ watch param) | Watch | WatchNamed | Collection: */ns/pods?watch=true*, One: */ns/pods/[name]?watch=true*, Subresource: */ns/pods/[name]/special?watch=true* |
| POST | Create | CreateNamed | Collection: */ns/pods*, One: */ns/pods/[name]*, SubResource: */ns/pods/[name]/generate* |
| PUT | UpdateCollection | Update | Collection */ns/pods*, One: */ns/pods/[name]*, SubResource: */ns/pods/[name]/status* |
| DELETE | DeleteCollection | Delete | Collection */ns/pods*, One: */ns/pods/[name]*, SubResource: */ns/pods/[name]/status* |

kbeecher commented:
@csrwng Thanks for the info. Is this only in planning so far, or do you also have some code/PRs that I can look at?

csrwng commented:
@kbeecher mostly planning. I submitted a PR to start that change with the Create method here: https://github.com/GoogleCloudPlatform/kubernetes/pull/7718 with corresponding comments in the rest package: https://github.com/csrwng/kubernetes/commit/fd65427e28dac5675a6c287e11eb581d537f6f10#diff-47bcf1e79133be8a14b98c919e1d1361R31

derekwaynecarr commented:
I don't see a reason to gate this work with what Cesar is looking at.  They seem aligned from what I can gather.  I would like to get this in and used before Kube 1.0

Sent from my iPhone

> On May 11, 2015, at 8:48 AM, Cesar Wong <notifications@github.com> wrote:
> 
> @kbeecher mostly planning. I submitted a PR to start that change with the Create method here: #7718 with corresponding comments in the rest package: csrwng@fd65427#diff-47bcf1e79133be8a14b98c919e1d1361R31
> 
> —
> Reply to this email directly or view it on GitHub.
>

derekwaynecarr commented:
Please poke the PR when its ready for another review pass.

kbeecher commented:
@derekwaynecarr I added a bit more if you'd like to make another pass. I've been trying to widen the change to cover more types of resources.

There are now DeleteAll methods in clients for Events, LimitRanges, PersistentVolumeClaims, Pods, Controllers, ResourceQuotas, Secrets and Services (i.e. all resources that are deleted en masse in the namespace_controller).

There are also DeleteCollection delegates for all resources you mentioned in an earlier comment (i.e. Events, Services, Controllers, ResourceQuotas, LimitRanges, Namespaces) plus Secrets and PersistentVolumeClaims.

I'm not sure if the remaining resource types should support this operation, but please let me know your thoughts.

derekwaynecarr commented on pkg/registry/event/registry.go:
```diff
@@ -47,6 +47,10 @@ func NewEtcdRegistry(h tools.EtcdHelper, ttl uint64) generic.Registry {
 			},
+			DeleteCollectionFunc: func(ctx api.Context) (runtime.Object, error) {
+				// TODO Create and return status object
+				return nil, h.Delete(etcdgeneric.NamespaceKeyRootFunc(ctx, prefix), true)
```
Can you define a common function that all namespaced resources can use as the ```DeleteCollectionFunc``.

For example:

```code
func DeleteCollectionPerNamespaceFunc func(...) ... {
  return &api.Status{...}, ...
}
```

derekwaynecarr replied:
In that common function, please make sure you return an error if there is no namespace in the provided context to ensure that we do not get into a situation where all resources across all namespaces are inadvertently deleted.

derekwaynecarr commented:
This is looking really close to being ready to merge.

Please make the update to have a common namespace deletion func, and ensure that that function throws an error if there is no namespace in the context.

Ping when that's ready, and I will do a final pass.

Thanks again!  Look forward to getting this in.

derekwaynecarr commented:
Looks like namespace_controller_test.go needs a minor update to note that it is now calling the deleteAll style methods as part of namespace termination.

kbeecher commented:
@derekwaynecarr I've pushed changes that add a DeleteCollectionPerNamespaceFunc and fix the namespace controller test.

There's just one more test that's failing (TestEtcdDeleteCollection), which I will have to look at tomorrow.

derekwaynecarr commented:
Will review. Please ping when tests are in.  Would like to get this in soon.  Thanks again. 

Sent from my iPhone

> On May 20, 2015, at 5:13 PM, Karl Beecher <notifications@github.com> wrote:
> 
> @derekwaynecarr I've pushed changes that add a DeleteCollectionPerNamespaceFunc and fix the namespace controller test.
> 
> There's just one more test that's failing (TestEtcdDeleteCollection), which I will have to look at tomorrow.
> 
> —
> Reply to this email directly or view it on GitHub.
>

kbeecher commented:
I fixed the test.

derekwaynecarr commented:
I will dedicate time for a review of this tomorrow.

derekwaynecarr commented:
Testing this now, would love to get this merged today.

Can you rebase and run ```hack/update-swagger-spec.sh``` to ensure that the swagger documentation is up to date to capture the change to DELETE at collection level?

derekwaynecarr commented:
Can you make sure service accounts also supports the deleteAll semantic?

derekwaynecarr commented:
Needs a rebase, I would really like to see this merged before feature freeze.

see #9197 for some related discussion on performance of deleting namespace with a large number of events that would be improved by this feature.  while core issue in #9197 was solved, the solution here is the longer term ideal.

It looks like graceful deletion of pods may or may not get in, need to synch with @smarterclayton tomorrow on how we should proceed here.

kbeecher commented:
@derekwaynecarr Rebased.

bgrant0607 commented:
FYI: #10217

ixdy commented:
Whenever this PR gets picked up again: please enable the `Namespaces Delete ALL of 100 namespace in 150 seconds` test in `test/e2e/namespaces.go`.

derekwaynecarr commented:
@kbeecher - I know this PR has been opened for a while, but now that Kubernetes 1.0 has been released, I am interested in getting this in ASAP.  Are you still able to rebase and update this PR, or would you like another member of community to pick up the work?

kbeecher commented:
@derekwaynecarr I can pick this up again in the next 1-2 days. I've already rebased and made everything build successfully again. I just need to update the unit tests.

wojtek-t commented:
@kbeecher - when do you expect this PR to be ready? I agree with @derekwaynecarr that we would like to have it in ASAP. We really suffer from "slow namespace deletion" in our e2e scalability tests.

kbeecher commented:
@derekwaynecarr @wojtek-t I've pushed the latest WIP. I've been trying to adapt everything to the latest changes over the last few weeks.

I think I've mostly succeeded, but there is one test which is failing for me. The function at `pkg/registry/generic/etcd/etcd.go#146` expects a runtime.Object but I don't know what to supply to it and my current solution causes the unit tests to fail.

Can anyone help?

derekwaynecarr commented:
I will take a look Monday to help out.  Thanks!

On Friday, July 31, 2015, Karl Beecher <notifications@github.com> wrote:

> @derekwaynecarr <https://github.com/derekwaynecarr> @wojtek-t
> <https://github.com/wojtek-t> I've pushed the latest WIP. I've been
> trying to adapt everything to the latest changes over the last few weeks.
>
> I think I've mostly succeeded, but there is one test which is failing for
> me. The function at pkg/registry/generic/etcd/etcd.go#146 expects a
> runtime.Object but I don't know what to supply to it and my current
> solution causes the unit tests to fail.
>
> Can anyone help?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/7372#issuecomment-126724930>
> .
>

wojtek-t commented:
@kbeecher - I looked into Shippable and Travis failures - it seems both are failing on something else:
"Please run hack/update-swagger-spec.sh"

kbeecher commented:
@wojtek-t I updated the swagger spec. Hopefully you will be able to see the error I refer to now.

wojtek-t commented on pkg/registry/generic/etcd/etcd.go:
```diff
@@ -131,6 +135,26 @@ func NamespaceKeyFunc(ctx api.Context, prefix string, name string) (string, erro
+		if !ok || len(ns) == 0 {
+			return nil, kubeerr.NewBadRequest("Namespace parameter required.")
+		}
+		var out runtime.Object
```
You should pass NewListFunc (from Etcd) to this function and use:
"out := NewListFunc()" 
instead.

wojtek-t commented:
@kbeecher - sorry for late response. I commented on that file how to solve this problem (hopefully it would help).

Also, please rebase.

kbeecher commented:
@wojtek-t Thanks for the response. I've rebased and refactored, but I must admit to being a bit confused by the errors I'm getting from the unit tests - I guess the upstream changes have altered some behaviour but I can't work it out.

derekwaynecarr commented:
I have time tomorrow to check out the unit tests.  Will look to get resolved.  Thanks for all you have done this far.

wojtek-t commented:
@kbeecher - regarding test failures:
- TestSyncNamespaceThatIsTerminating is the test problem - you need to fix the test
- TestEtcdDeleteCollection - it seems that you're not registring DeleteCollectionFunc for pods.

derekwaynecarr commented:
Hacking on a fix now, will push a link to a branch with the additive changes for you tor review when done.

derekwaynecarr committed:
Modify REST framework to support deleting all resources in a namespace

derekwaynecarr commented:
@kbeecher - I think I have the necessary changes here:

https://github.com/derekwaynecarr/kubernetes/commit/2d08d6699b92a0e3807a0bde01a7b3e85f0aa562

Please take a look.  Unit tests pass.  Need to test e2e still, but you should be unblocked with this.

derekwaynecarr commented:
Hmm, so it looks like we are not registering the DELETE of a collection properly now with api server.  

Will update my commit branch, and ping the PR when that is working for you to review.

derekwaynecarr commented:
@kbeecher - I have to end for the day, but I made forward progress here.

See https://github.com/derekwaynecarr/kubernetes/commit/ed57f2ccebfaf331878b39e97bdaa7ecf138fa77

It has the proper registration with the server now, unit tests pass, but we are missing DeleteCollection implementations for services and events.  I was able to confirm that I was able to delete the replication controller collection before running out of time.

Let me know if you want me to finish up this PR or if you are good to go from here.

wojtek-t commented:
Regarding services - see also #12340 which is changing the services registry.

derekwaynecarr commented:
@wojtek-t - that makes things easier, will hold updating this until #12340 merges.

wojtek-t commented:
@derekwaynecarr - thanks for FYI (now I know it's blocking something, although I'm now just waiting for review)

wojtek-t commented:
@derekwaynecarr - #12340 that you were waiting for is already merged.

derekwaynecarr commented:
Time permitting, I will iterate on this a little more.  Thanks for the
update.

On Tuesday, August 11, 2015, Wojciech Tyczynski <notifications@github.com>
wrote:

> @derekwaynecarr <https://github.com/derekwaynecarr> - #12340
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/12340> that you
> were waiting for is already merged.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/7372#issuecomment-129718726>
> .
>

kbeecher commented:
@derekwaynecarr I added the DeleteCollection function for services. I saw that events had already had one added.

derekwaynecarr commented:
@wojtek-t - it looks like in https://github.com/kubernetes/kubernetes/pull/12584 you removed the RecursiveDelete function.  This was used to delete the etcd directory in this PR.  I think calling the normal delete call that takes a ```out runtime.Object``` is problematic since we have no object that represents a directory.  

Which do you prefer:

1. Add a delete collection operation that does what recursive delete did, but requires the input key to be a dir.
2. Call the existing delete, but update to ignore ```out``` if its nil in the storage.Delete call or if the prevNode was a directory?

wojtek-t commented:
Ooops - sorry for that.

I think 1 is probably better (but I'm not against 2) - so I think whatever is easier for you is fine.

derekwaynecarr commented:
I will hack on doing 1

wojtek-t commented:
@derekwaynecarr Do you need some help with this PR?

derekwaynecarr commented:
@wojtek-t - I think I should be able to push a branch that is working for @kbeecher to diff from today.  I am on vacation next week, so if this PR continues to linger then it would be great for you to help assist.  This PR has been open so long it's major surgery to bring it back to life.

wojtek-t commented:
@derekwaynecarr - thanks - please push whatever you have at the end of the day; I can take it from Monday

derekwaynecarr commented:
@wojtek-t - new issue I am running into is related to behavior of the new etcd_watcher.go and how it responds to a notification of a directory node being deleted.

If you look here:
https://github.com/kubernetes/kubernetes/blob/master/pkg/storage/etcd/etcd_watcher.go#L333

We make an assumption that the removed object is never a directory.

This appears to cause the api server to eventually crash and get overwhelmed when running with this current code.

Aside from that issue, I think we should not migrate Services to use DeleteCollection yet because there is too much logic to delete related endpoints and release any relevant ips.

Anyway, thoughts on how we should handle etcd_watcher in response to a directory delete?

derekwaynecarr commented:
@wojtek-t - i think there was an unrelated issue that was causing haywire, i may have reported a problem too soon.

derekwaynecarr commented:
@wojtek-t - the code is here:

https://github.com/derekwaynecarr/kubernetes/tree/namespace_delete

The major problem is that deleting a directory node does not seem to cascade events properly to watchers.  

This is best illustrated in the following scenario:

```
kubectl create -f docs/admin/namespaces/namespace-dev.json
kubectl create -f docs/user-guide/resourcequota/quota.yaml --namespace=development
kubectl create -f docs/admin/limitrange/limits.yaml --namespace=development
kubectl run nginx --image=nginx --replicas=2 --namespace=development
kubectl delete namespaces/development

everything is deleted, but when I attempt to re-use the namespace again...

kubectl create -f docs/admin/namespaces/namespace-dev.json
kubectl run nginx --image=nginx --replicas=100 --namespace=development
Error from server: ReplicationController "nginx" is forbidden: resourcequotas "quota" not found
```

The issue is that our admission controllers are not getting watch notifications for the deleted objects.  If I restart kube-apiserver, I am able to work again with that namespace.

Aside from that major issue, services are still deleted one by one, which means we still have a lingering directory in etcd when the namespace is deleted.  We need to think through the best way to do this in bulk so that we can do the clean-up of endpoints and release ips.

derekwaynecarr commented:
@wojtek-t - this is frustrating, but it seems like our only option is to special case our proxy on top of etcd_watcher.go to understand how to invalidate keys below a specific path.  It looks to me from testing etcd that you only get notified of the parent directory being deleted, and you are unable to know what children were in that node.  It would have been awesome if we could have got the first level children keys even if not their content.  I am not sure how to proceed here as this seems like a pretty fundamental problem with using directory scoped deletion in etcd.

derekwaynecarr commented:
@wojtek-t 

to demonstrate:

I created following in etcd:

/pods/ns1/a
/pods/ns1/b

Set the watch on /pods
Delete /pods/ns1?recursive=true
And watch gets only following information:

```
curl http://localhost:4001/v2/keys/pods?recursive=true\&wait=true
{"action":"delete","node":{"key":"/pods/ns1","dir":true,"modifiedIndex":9,"createdIndex":7},"prevNode":{"key":"/pods/ns1","dir":true,"modifiedIndex":7,"createdIndex":7}}
```

If etcd could give us the first level children in /pods/ns1 we could have done more, but we get nothing of the sort for directory objects.

wojtek-t commented:
>
@wojtek-t - this is frustrating, but it seems like our only option is to special case our proxy on top of etcd_watcher.go to understand how to invalidate keys below a specific path. It looks to me from testing etcd that you only get notified of the parent directory being deleted, and you are unable to know what children were in that node. It would have been awesome if we could have got the first level children keys even if not their content. I am not sure how to proceed here as this seems like a pretty fundamental problem with using directory scoped deletion in etcd.

I've got your point - yes seems that currently we can also extend etcd_watcher.go (btw thanks a lot for investigating it)

@xiang90 @yichengq
 Do you think that it makes sense to add a support "remove the whole directory" in the watch at etcd level? Currently (as described above by @derekwaynecarr) if you remove the whole directory, you will not get notification about individual objects via watch. It would be great if you could add a support for it to etcd. Thoughts?

kbeecher commented:
@derekwaynecarr I will soon be going on parental leave for several weeks and so will have little or no time to work on this. Should I hand this work over to someone?

wojtek-t commented:
@xiang90 @yichengq
Do you think that it makes sense to add a support "remove the whole directory" in the watch at etcd level? Currently (as described above by @derekwaynecarr) if you remove the whole directory, you will not get notification about individual objects via watch. It would be great if you could add a support for it to etcd. Thoughts?

xiang90 commented:
@wojtek-t We do not expect any further changes for the v2 API if it is not a hard requirement. 

I am thinking how we should design the watching in v3. Seems like some people want to see the actual changes(a dir and keys under it are removed), but some want to know about the rough action (a dir is removed).

wojtek-t commented:
>
@wojtek-t We do not expect any further changes for the v2 API if it is not a hard requirement.

Fair enough - thanks for the answer.

>
I am thinking how we should design the watching in v3. Seems like some people want to see the actual changes(a dir and keys under it are removed), but some want to know about the rough action (a dir is removed).

My personal feeling is that if you are observing say "a/b/c/d" element (or directory) you would like to see that "a/b/c/d" was deleted, not that some of it's parent (e.g. "a/b") was deleted. What do you think?

xiang90 commented:
> My personal feeling is that if you are observing say "a/b/c/d" element (or directory) you would like to see that "a/b/c/d" was deleted, not that some of it's parent (e.g. "a/b") was deleted. What do you think?

In v2, we return the action not the changes that is why you observed the directory is removed. In v3, we do not even have the concept of a dir. If you do a recursive removal, it can send ALL items removed to a watcher or just the recursive removal action. If we choose one, then there might be a huge overhead. We might want to add an additional option to watch. I am still not sure. Glad you mentioned it, we need to keep this in mind.

derekwaynecarr commented:
@wojtek-t - did you have any ideas on how to proceed here, the goal still remains to more efficiently purge resources, wondering if you thought of any fancy way to get around the problem we encountered.

wojtek-t commented:
@derekwaynecarr - sorry I didn't have much time to think about it.

wojtek-t commented:
@derekwaynecarr - I'm getting back to it since I would really like to make it happen.

I think that we can solve the watch problem using the the "watch in apiserver" mechanism that I added some time ago.
Basically, the cacher is storing the current state of resources in memory. So it is enough for it to observe just an action of "directory deleted", because it can iterate over its elements and remove all that were in that directory.
That would require switching it on for all resource, but we wanted to do it anyway.
WDYT?

However, I think that there might be one more problem - basically I'm talking about graceful deletion. What should happen if DeletionGracePeriodSeconds is set for an object and we are removing directory that contains such objects? I think that shouldn't be ignored... @smarterclayton

derekwaynecarr commented:
We had discussed leaving pod deletion to happen on a per item basis, but
moving all other deletions to a collection level delete.

Do you have an idea for how much complication this adds to our watch cache
to understand hierarchy in this manner?

On Thursday, October 8, 2015, Wojciech Tyczynski <notifications@github.com>
wrote:

> @derekwaynecarr <https://github.com/derekwaynecarr> - I'm getting back to
> it since I would really like to make it happen.
>
> I think that we can solve the watch problem using the the "watch in
> apiserver" mechanism that I added some time ago.
> Basically, the cacher is storing the current state of resources in memory.
> So it is enough for it to observe just an action of "directory deleted",
> because it can iterate over its elements and remove all that were in that
> directory.
> That would require switching it on for all resource, but we wanted to do
> it anyway.
> WDYT?
>
> However, I think that there might be one more problem - basically I'm
> talking about graceful deletion. What should happen if
> DeletionGracePeriodSeconds is set for an object and we are removing
> directory that contains such objects? I think that shouldn't be ignored...
> @smarterclayton <https://github.com/smarterclayton>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/7372#issuecomment-146461547>
> .
>

wojtek-t commented:
>
We had discussed leaving pod deletion to happen on a per item basis, but
moving all other deletions to a collection level delete.

So do we implicitly assume that graceful deletion makes sense only for pods? The graceful deletion period is defined at the ObjectMeta level, so potentially it can be set for any object.
I'm not saying this is a wrong approach, but if so we need to document it well.

>
Do you have an idea for how much complication this adds to our watch cache
to understand hierarchy in this manner?

Actually - I just looked into the code and I'm afraid I was wrong :(
In the Cacher we explicitly get an object that was deleted (it's already processed by etcd_watcher.go).
So we would have to modify etcd_watcher to understand it and also modify watch interface to enable passing such values (cacher in fact is a decorator of etcd_helper).

So unfortunately it will NOT be easy...

smarterclayton commented:
Graceful is going to extend over other resources in the future on a case by
case basis.  Pods and namespaces are the only concrete gracefuls, and
namespaces have extra semantics that possibly should be aligned in the
future.

I suspect bulk deletion is orthogonal to whether you can graceful
deletion.  If we have bulk delete with grace, doing a bulk set is
reasonable.  I don't see an easy way to make that efficient though.  It's
worth noting that the Kubelet handles a raw delete with no graceful delete
metadata set as "use the default value from the pod spec", which means if
you implement bulk delete as is you're getting "bulk delete with graceful
termination on the Kubelet".  We could reject delete calls with explicit
grace periods in the short term.

On Oct 8, 2015, at 8:53 AM, Wojciech Tyczynski <notifications@github.com>
wrote:

We had discussed leaving pod deletion to happen on a per item basis, but
moving all other deletions to a collection level delete.

So do we implicitly assume that graceful deletion makes sense only for
pods? The graceful deletion period is defined at the ObjectMeta level, so
potentially it can be set for any object.
I'm not saying this is a wrong approach, but if so we need to document it
well.

Do you have an idea for how much complication this adds to our watch cache
to understand hierarchy in this manner?

Actually - I just looked into the code and I'm afraid I was wrong :(
In the Cacher we explicitly get an object that was deleted (it's already
processed by etcd_watcher.go).
So we would have to modify etcd_watcher to understand it and also modify
watch interface to enable passing such values (cacher in fact is a
decorator of etcd_helper).

So unfortunately it will NOT be easy...

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/7372#issuecomment-146530887>.

wojtek-t commented:
@derekwaynecarr - I talked to coreos folks about the watch issue and they already solve it in v3 api, but are not going to solve it in v2 api - see https://github.com/coreos/etcd/issues/3672 for reference.
So I think that we should wait until we will be using v3 api in k8s and solve it then - WDYT?

wojtek-t commented:
@derekwaynecarr @lavalamp @gmarek 
I thought about it one more time, and I think there is a way to proceed with this PR.
So what I would like to do is to:
- implement DeleteCollection in apiserver as: "List + Delete all returned items one by one"
- [this is obviously not atomic, but for deleting namespace is good enough]
- once we migrate to v3 api in etcd, we can simply get rid of this code in apiserver and use appropriate etcd api call
- however, writing this additional code in apiserver should be extremely simple and is not much overhead

The advantage of it is that deleting namespace can be much much faster, because we won't be blocked on namespace-controller QPS limit (and this will help us a lot in scalability tests)

Any thoughts on it? If you agree with it, I'm happy to work on it next week.

derekwaynecarr commented:
@wojtek-t - so if I understand correctly, we will still do N deletes to etcd, but client will make a single call to the API server?  That sounds fine to me, fixes the watch problem for time being.  I didn't think the v3 model fixed our watch problem either.  I assume if request times out, caller just needs to know to try again (and the remaining items will be deleted)?

wojtek-t commented:
>
@wojtek-t - so if I understand correctly, we will still do N deletes to etcd, but client will make a single call to the API server? 

That's what I'm suggesting.

>
That sounds fine to me, fixes the watch problem for time being. I didn't think the v3 model fixed our watch problem either. 

In v3 api, this problem is solved - I talked to coreos guys about it.

>
I assume if request times out, caller just needs to know to try again (and the remaining items will be deleted)?

Yes - I guess so.

derekwaynecarr commented:
@wojtek-t - sgtm.

lavalamp commented:
I'm sorry for being so late to this party, but can we make the collection-level deletion take the same parameters as list (and watch)--e.g., label and field selectors--and *not* call it "delete-all"? (maybe "bulk-delete") I think that'd be much more generally useful!

lavalamp commented:
Yeah, I think @wojtek-t and I are on the same page.

derekwaynecarr commented:
please cc me on the resultant PR, i am interested in understanding how large the impact will be.

please confirm that the delete collection call also delete the directory node in the registry, and not just its child items as this PR intended?

wojtek-t commented:
>
please cc me on the resultant PR, i am interested in understanding how large the impact will be.

sure

>
please confirm that the delete collection call also delete the directory node in the registry, and not just its child items as this PR intended?

I will check if we can easily do it without v3 etcd api (I talked with coreos guys and it will definitely be possible in v3).

smarterclayton commented:
I'd prefer DELETE against the collection - we probably to support delete
options AND list options together, because otherwise there is no bulk
graceful delete (what namespace controller needs, vs what tear down needs).

On Dec 4, 2015, at 8:17 PM, Daniel Smith <notifications@github.com> wrote:

I'm sorry for being so late to this party, but can we make the
collection-level deletion take the same parameters as list (and
watch)--e.g., label and field selectors--and *not* call it "delete-all"?
(maybe "bulk-delete") I think that'd be much more generally useful!

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/7372#issuecomment-162122049>.

wojtek-t commented:
@smarterclayton - I agree that passing both ListOption & DeletionOptions is exactly what we need.

wojtek-t commented:
@derekwaynecarr  @smarterclayton - I've just send out the initial version of my idea in #18290 out for review.
Since this PR is even pretty hard to rebase, I think we should just close it.

lavalamp commented:
Yes, that SGTM

On Sat, Dec 5, 2015 at 7:45 AM, Clayton Coleman <notifications@github.com>
wrote:

> I'd prefer DELETE against the collection - we probably to support delete
> options AND list options together, because otherwise there is no bulk
> graceful delete (what namespace controller needs, vs what tear down needs).
>
> On Dec 4, 2015, at 8:17 PM, Daniel Smith <notifications@github.com> wrote:
>
> I'm sorry for being so late to this party, but can we make the
> collection-level deletion take the same parameters as list (and
> watch)--e.g., label and field selectors--and *not* call it "delete-all"?
> (maybe "bulk-delete") I think that'd be much more generally useful!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/7372#issuecomment-162122049
> >.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/7372#issuecomment-162213109>
> .
>

timothysc commented:
+1
